fem/bilininteg_mass_pa.cpp

```diff
@@ -51,27 +51,40 @@ void MassIntegrator::SetupPA(const FiniteElementSpace &fes, const bool force)
    dofs1D = maps->ndof;
    quad1D = maps->nqpt;
    pa_data.SetSize(ne*nq, Device::GetDeviceMemoryType());
-   Vector coeff;
+   printf("nq sz %d \n", nq);
+   Vector *coeff{nullptr};
+   bool own_coeff{true};
    if (Q == nullptr)
    {
-      coeff.SetSize(1);
-      coeff(0) = 1.0;
+      coeff = new Vector;
+      coeff->SetSize(1);
+      (*coeff)(0) = 1.0;
    }
    else if (ConstantCoefficient* cQ = dynamic_cast<ConstantCoefficient*>(Q))
    {
-      coeff.SetSize(1);
-      coeff(0) = cQ->constant;
+      coeff = new Vector;
+      coeff->SetSize(1);
+      (*coeff)(0) = cQ->constant;
+   }
+   else if (QuadratureFunctionCoefficient* cQ =
+               dynamic_cast<QuadratureFunctionCoefficient*>(Q))
+   {
+      const QuadratureFunction &qFun = cQ->GetQuadFunction();
+      coeff = &dynamic_cast<Vector &>(const_cast<QuadratureFunction &>(qFun));
+      own_coeff = false;
```

Comment:
@v-dobrev, @rcarson3 here we can just pull the vector with user defined scaling terms. Somehow though we would have to enforce that the data is indeed data at the quadrature point level. What do you guys think? If we go through eval that would be slow since it does it point wise.

Reply:
Yes, we do not want to call `Eval` to copy the data. 😄 

As a check, for now, it is probably good enough to just verify that the size of the quadrature data is as expected.

Instead of switching `coeff` to a pointer and adding `own_coeff`, you can just use `Vector::MakeRef` to make it point to the data in the `QuadratureFunction` without taking ownership. Note that before the `MakeRef` call, you will want to call `qFun.Read()` to make sure the data is available on the "mfem device" before making the alias memory with `MakeRef`.

Reply:
Also a good idea, will do.

Reply:
@rcarson3 , @v-dobrev , I added an MFEM_VERIFY to check that the quadraturefunctions are the expected size, but would there be a way to verify that the integration rules match?

Reply:
@artv3 before we removed the field interpolant  class in my last PR I verified the integration rules like so https://github.com/mfem/mfem/blob/quadrature-project-dev/fem/field_interpolant.cpp#L75-L98.  Maybe you could do something like that here?

Reply:
If you want to add a check, it is best to make it in a method of class `QuadratureSpace` -- just so we can reuse it in different integrators.

Probably this whole section that "converts" a `Coefficient` to a `Vector` should be turned into a protected method in the base class which can then be reused by all integrators.

Reply:
Another (maybe a better) option is to make this a virtual method in the class `Coefficient` -- this way there will be no need to do `dynamic_cast`s.